src/res/heatmap.html

```diff
@@ -0,0 +1,4294 @@
+
+	const zoomToGroupSize = [
+		1,              // 20 ms
+		50,             // 1 s
+		50 * 5,         // 5 s
+	];
+	const zoomToHeight = [
+		50,
+		60,
+		60,
+	];
+	const zoomToText = [
+		' 1 sec : 20 ms ',
+		' 1 min : 1 sec ',
+		' 5 min : 5 sec ',
+		' 1 hr  : 1 min '
+	];
+
+	const executionsHeatmap = new HeatmapCollection(claimHtml("executionsHeatmap"));
+	let cpool = [/*cpool:*/];
+	const methods = readMethods(new DataBuffer(claimHtml("methods")));
+	cpool = null;
+
+	let currentHeatmap = executionsHeatmap;
+	let currentHeatmapType = 'executions';
+
+	let heatLastSample = -1;
+	let heatActiveSample1 = -1;
+	let heatActiveSample2 = -1;
+	let heatDiffSample1 = -1;
+	let heatDiffSample2 = -1;
+	let highlightStart = -1;
+	let highlightEnd = -1;
+
+	const titleText = document.getElementById("title").textContent;
+	document.getElementById("title").style.display = 'none';
+	document.title = titleText;
+
+	const heatCanvas = document.getElementById('heatmap-canvas');
+	const heatStatus = document.getElementById('status');
+	const heatCanvasWrapper = document.getElementById('heatmap-canvas-wrapper');
+	const heatCanvasContainer = document.getElementById('heatmap-canvas-container');
+
+	let heatC;
+	let prevDx = -1000000000;
+
+	const timeOptions = {
+		year: '2-digit',
+		month: '2-digit',
+		day: '2-digit',
+		timeZoneName: 'short',
+		hour: '2-digit',
+		minute: '2-digit',
+		second: '2-digit',
+		fractionalSecondDigits: 2
+	};
+
+	const timeOptionsShort = {
+		hour: '2-digit',
+		minute: '2-digit',
+		second: '2-digit',
+		fractionalSecondDigits: 2
+	};
+
+	const timeOptionsTiny = {
+		hour: '2-digit',
+		minute: '2-digit',
+		second: '2-digit',
+		hourCycle: 'h23',
+		//hour12: false
+	};
+
+	function getCount(counts, i) {
+		if (i < 0 || counts.length === 0) {
+			return 0;
+		}
+		return counts[Math.min(i, counts.length - 1)];
+	}
+
+	let bgInterval = null;
+	let bgTasks = [];
+	let cooldownTime = 0;
+	let nextTask = 0;
+
+	function addTask(id, iteration, afterFrame) {
+		for (let t of bgTasks) {
+			if (t.id === id) {
+				t.f = iteration;
+				t.a = afterFrame;
+				return;
+			}
+		}
+
+		bgTasks.push({id: id, f: iteration, a: afterFrame});
+		if (bgTasks.length === 1) {
+			bgInterval = setInterval(function () {
+				let start = performance.now();
+				if (start <= cooldownTime) {
+					return;
+				}
+				let overflow = false;
+				do {
+					let task = bgTasks[nextTask];
+					let runAgain;
+					try {
+						runAgain = task.f();
+					} catch (e) {
+						console.error(e);
+						bgTasks.length = 0;
+						clearInterval(bgInterval);
+						nextTask = 0;
+						return;
+					}
+
+					if (runAgain) {
+						nextTask++;
+					} else {
+						bgTasks.splice(nextTask, 1);
+						if (bgTasks.length === 0) {
+							clearInterval(bgInterval);
+							nextTask = 0;
+							return;
+						}
+					}
+
+					if (nextTask >= bgTasks.length) {
+						nextTask = 0;
+						overflow = true;
+					}
+				} while (performance.now() - start < 8);
```

Comment:
Just wondering about this magic value

Reply:
Numbers from 4 to 20 work nice)))
Big numbers make UI stuck, low numbers make calculations slow.

Reply:
I'll make some named constant

Reply:
fixed